File: templates/r/boxplot.R
## 箱线图 + 抖动点 + 组间 p —— 卡片 02_组间比较图/箱线图_Boxplot.md
## 合成数据，可独立运行： Rscript templates/r/boxplot.R
.args <- commandArgs(trailingOnly = FALSE)
.sd <- dirname(normalizePath(sub("^--file=", "", .args[grep("^--file=", .args)])))
source(file.path(.sd, "_common.R"))
suppressPackageStartupMessages({ library(ggpubr); library(ggbeeswarm) })
set.seed(2026)

# >>> PARAM ------------------------------------------------------
NAME <- "tpl_boxplot"; TITLE <- "Boxplot + jitter + p"; XLAB <- NULL; YLAB <- "Value"
PALETTE <- pal("med_case_control")            # 二分类（高=暖 低=冷）
# 换真实数据： df <- readr::read_csv("data.csv"); 列 grp / val
df <- data.frame(grp = factor(rep(c("Group 1", "Group 2"), each = 150)),
                 val = c(rnorm(150, 0, 1), rnorm(150, .6, 1.1)))
# <<< PARAM ------------------------------------------------------

cols <- c("Group 1" = unname(PALETTE["low"]), "Group 2" = unname(PALETTE["high"]))
p <- ggplot(df, aes(grp, val, fill = grp)) +
  geom_boxplot(outlier.shape = NA, width = .55, alpha = .85) +
  geom_quasirandom(size = .4, alpha = .15, color = "grey25") +
  stat_compare_means(method = "wilcox.test", label = "p.format", label.x = 1.3) +
  scale_fill_manual(values = cols) +
  labs(x = XLAB, y = YLAB, title = TITLE) + theme(legend.position = "none")
save_gg(p, NAME)

File: templates/python/boxplot.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""箱线图 + 抖动点 —— 卡片 02_图表类型知识卡片/02_组间比较图/箱线图_Boxplot.md
合成数据，可独立运行。改 # >>> PARAM 区换数据/列/配色。"""
import os, sys, warnings; warnings.filterwarnings("ignore")
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np, pandas as pd, seaborn as sns, matplotlib.pyplot as plt
from _common import set_theme, save_fig, pal
set_theme(); np.random.seed(7)

# >>> PARAM ------------------------------------------------------
NAME = "tpl_boxplot"
TITLE = "Boxplot + jittered points"; XLAB = None; YLAB = "Value"
PALETTE = "cat_main"          # 分类配色 registry 名
FIGSIZE = (3.4, 2.9)
# 换真实数据： df = pd.read_csv("data.csv"); X="grp"; Y="val"
X, Y = "grp", "val"
N = 200
df = pd.DataFrame({X: np.repeat(["A", "B", "C"], N),
                   Y: np.concatenate([np.random.normal(0, 1, N),
                                      np.random.normal(.8, 1.2, N),
                                      np.random.normal(.3, .8, N)])})
# <<< PARAM ------------------------------------------------------

colors = pal(PALETTE)
fig, ax = plt.subplots(figsize=FIGSIZE)
sns.boxplot(data=df, x=X, y=Y, hue=X, palette=colors[:df[X].nunique()],
            legend=False, showfliers=False, width=.55, ax=ax)
sns.stripplot(data=df, x=X, y=Y, color="#333333", size=2, alpha=.25, ax=ax)
ax.set_title(TITLE); ax.set_xlabel(XLAB); ax.set_ylabel(YLAB)
save_fig(fig, NAME)
